271135 visual studio2017port

afdeprado opened the pull request:
This is the initial work on porting MuseScore to the Visual Studio 2017 toolchain.

The current status: the project is using the CMake build system, and creates the solution and projects for Visual Studio 2017. The solution compiles and links with no errors, but with lots of warnings:
Project awl: 8 warnings
Project bww: 1 warning
Project libmscore: 259 warnings
Project mscore: 330 warnings
Project poppler: 660 warnings
All the other projects compile with no warnings.
When trying to execute the resulting program, an exception is triggered early in the initialization process (line 343 - libmscore/style.cpp, the QVariant::fromValue<Direction>() method is throwing an exception. All QVariant::fromValue() invocations for the enum Align work flawlessly. The only difference I've noticed is that for Ms::Align, Q_DECLARE_METATYPE(Ms::Align) is invoked in elementlayout.h, whereas for Direction this is not used, and Q_ENUM_NS(Direction) is. Could this be the source of the problem? ).

I have created separately a document with the build instructions for Visual Studio 2017, as well as a file containing the required dependencies, which I'm including here, The build instructions for VS2017 should be published alongside the instructions for other platforms.

[dependencies.zip](https://github.com/musescore/MuseScore/files/1945534/dependencies.zip)
[Compile Instructions MuseScore VS2017.docx](https://github.com/musescore/MuseScore/files/1945583/Compile.Instructions.MuseScore.VS2017.docx)

Jojo-Schmitz commented:
Any special reason for the many renamend variables, is this needed vor VS?
Also the changes of foreach( ; ) to for( : ), are they really needed for VS?
And more than 100 commits is way too much I believe.

afdeprado commented:
Hi!

- Renamed variables: this is not strictly necessary. I set the project to compile with warning level 4 (/W4), which is akin but more stringent than -Wall -Wextra. This generates lots of warnings for issues that might lead to bugs. Among the flagged issues are name hiding warnings, when a variable created in an inner scope hides a variable in an outer scope. The variable renamings are to get rid of these warnings. This is also consistent with the recommendation ES.12 in the C++ Core Guidelines by Stroustrop & Sutter (http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-reuse : ES.12: Do not reuse names in nested scopes).
- foreach(,) -> for(:): a few of these generate warnings due to name hiding (as the macro expansion of foreach() uses always the same local variable name). Most I changed as per the "MuseScore Coding Rules" (https://musescore.org/en/handbook/musescore-coding-rules#Loops), which tells explicitly "Use C++11's "for" instead of Qt's "foreach":" ... "If you happen to be fixing some code and see a "foreach", please change that loop into a "for"."
- The real number of commits is more like 35, but I did 2 rebases (one last week, one yesterday), and each one duplicated the number of commits (it must have been a mistake I made, I'm new to git). I'm looking into how to squash the commits, but I'm worried about losing my work if I don't do it right; squashing seems to be a very unforgiving operation... But I would definitely want to have all squashed to a single commit.
I'll try to join the IRC channel later on; perhaps we can talk this over.
Regards!

anatoly-os commented:
Thank you for input! You probably know that compiling master with MSVC will allow using Start Center with Qt Webengine.

Btw, there is a good practice to put only the changes related to the PR topic, not more. Making too many changes lead to messed history and blame. You are welcome to create separate PRs with renaming and other stuff.

anatoly-os commented:
To avoid so many commits, you can use squashing commits with `git rebase -i HEAD~XX`, where XX is a number of commits you want to change. And then use `squash` instead of `pick` for all the commits you want to squash keeping the valuable commit log for the one left commit.

Jojo-Schmitz commented:
Just had a chat with @afdeprado , all the variable renamings and the foreach changes are to avoid compiler warnings. So I'd like to have one commit with the changes to get it to build with VS and another to deal with the warnings, or something like this

afdeprado commented:
As discussed with @Jojo-Schmitz, compacted the commits and separated issues. There are now 4 commits:
- 64515da: Changes to the build system (CMake files), and added files necessary to build under MSVC (in particular, a LGPL'd getopt() port for MSVC),
- 67162af: Changes to compile with no errors on MSVC.
- c4d2a5b: Changes to eliminate warnings. This is WIP, as there are still 1257 unsolved warnings (see detail on first post). 
- 4636a77: Post-rebase added change to build system (one remaining problem).

I have rebased, and compiled before force-pushing the changes, to make sure there are no new errors.

Build instructions and dependencies for MSVC are attached to the first post in this conversation. I will start a posting in the forum about the port to MSVC.

Jojo-Schmitz commented on awl/aslider.cpp:
```diff
@@ -302,7 +302,10 @@ QString AccessibleAbstractSlider::text(QAccessible::Text t) const
                   return QString();
             }
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+      // Eliminate "unreachable code" warning for VS compile target.
```
If that code is unreachable irrespectively of compiler used, remove it altogether

Jojo-Schmitz commented on libmscore/barline.cpp:
```diff
@@ -566,7 +566,10 @@ printf("may drop\n");
                && segment()->isEndBarLineType());
             }
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+      // Prevent unreachable code warning in MSVC
```
If that code is unreachable irrespectively of compiler used, remove it altogether

Jojo-Schmitz commented on libmscore/chordrest.cpp:
```diff
@@ -843,7 +843,10 @@ QVariant ChordRest::propertyDefault(Pid propertyId) const
                   return DurationElement::propertyDefault(propertyId);
             }
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+      // Prevent unreachable code warning in MSVC
```
If that code is unreachable irrespectively of compiler used, remove it altogether

Jojo-Schmitz commented on midi/midifile.cpp:
```diff
@@ -293,7 +293,10 @@ bool MidiFile::read(QIODevice* in)
             default:
                   throw(QString("midi file format %1 not implemented").arg(_format));
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+                  // Prevent "unreachable code" warning in MSVC
```
If that code is unreachable irrespectively of compiler used, remove it altogether

afdeprado commented:
I will review the changes and push a new commit with them.

anatoly-os commented:
If it works, it is really cool! I will try to build your PR tomorrow and check whether Webengine works in 3.0

afdeprado commented:
Anatoly, right now, it *compiles* with no errors... but it doesn't execute successfully. 

The first issue I encountered, as described above, is that an exception is triggered early in the initialization process (line 343 - libmscore/style.cpp, the QVariant::fromValue()). Yesterday, I was able to solve this (not published yet), by adding Q_DECLARE_METATYPE(Ms::Direction) in libmscore/types.h, after the declaration of the Direction enum.

After this, a new exception is thrown later on, but I haven't analyzed it yet...

anatoly-os commented:
Ok, got it. If I have time, I will have a look at the issues, too. Great thanks to your contribution, again :)

afdeprado commented on thirdparty/portmidi/pm_common/portmidi.c:
```diff
@@ -154,40 +164,73 @@ const PmDeviceInfo* Pm_GetDeviceInfo( PmDeviceID id ) {
 PmError pm_success_fn(PmInternal *midi) {
-    return pmNoError;
+#if (defined (_MSCVER) || defined (_MSC_VER))
+   UNREFERENCED_PARAMETER(midi);
```
portmidi.c is *not* including Qt headers, and it is a plain C (not C++) file. Q_UNUSED() is unavailable, unless I copy the definition from the relevant header file. I propose to change the definition of UNREFERENCED_PARAMETER(P) to define as-is for MSVC and as a no-op (replace with nothing) for other platforms, and remove the #if conditionals throughout the code. Does this sound good?

afdeprado commented on thirdparty/portmidi/pm_common/portmidi.c:
```diff
@@ -6,6 +6,16 @@
 
+#if (defined (_MSCVER) || defined (_MSC_VER))
+/* Define UNREFERENCED_PARAMETER macro to suppress warnings about unreferenced parameters. */
+#define UNREFERENCED_PARAMETER(P)          (P)
```
(And also regarding previous one): UNREFERENCED_PARAMETER is a macro defined in one of the MSVC headers, which isn't included by portmidi.c, therefore I just copied the definition. And I didn't know Q_UNUSED(), will use that definitely!

afdeprado replied:
portmidi.c is not including Qt headers, and it is a plain C (not C++) file. Q_UNUSED() is unavailable, unless I copy the definition from the relevant header file. I propose to change the definition of UNREFERENCED_PARAMETER(P) to define as-is for MSVC and as a no-op (replace with nothing) for other platforms, and remove the #if conditionals throughout the code. Does this sound good?

afdeprado commented:
Revised commits:

f3c8931: Changes to the build system (CMake files), and added files necessary to build under MSVC (in particular, a LGPL'd getopt() port for MSVC),
f42240a: Changes to compile with no errors on MSVC.
9e0ddb8: Changes to eliminate warnings in MuseScore code. This is WIP, as there are still 1257 unsolved warnings (see detail on first post).
b766415: Changes to eliminate warnings in 3rd party code. These changes could be initially discarded, to avoid modifying 3rd-party code.
I have rebased again to keep it up-to-date..

Main changes from previous commits:
- Unreachable code is now commented unconditionally
- Q_UNUSED() instead of UNREFERENCED_PARAMETER(), except in a couple of 3rd-party files (in portmidi), which are Windows-specific, and which already include winnt.h, where UNREFERENCED_PARAMETER() is defined.
- Var initialization is now done unconditionally for all compilers.

Build instructions and dependencies for MSVC are attached to the first post in this conversation. I will start a posting in the forum about the port to MSVC.

Jojo-Schmitz commented:
Rebase needed.
And see also https://musescore.org/en/node/273288

Jojo-Schmitz commented:
I guess with #3796 this PR here can get closed?
